### ISSUE TITLE
Question on `eval_dex.py` `prev_state` handling vs paper offline protocol (benchmark.tex:121)

### ISSUE THREAD

**Issue by rakhimovv** [showlab/showui-pi#6]
Hi authors — first, thank you for the wonderful work on ShowUI-π. The unified discrete/continuous flow-matching formulation, the ScreenDrag benchmark with its data-driven online protocol, and the open release of the model + dataset + training code have been a great resource — really appreciated.

I've been working on reproducing the Tab 1 numbers as a baseline for some follow-up work, and noticed something in `scripts/eval_dex.py` that I wanted to ask about before drawing any conclusions. Happy to be corrected if I'm reading the protocol wrong.

### What I noticed

In `scripts/eval_dex.py:run_episode`, the state input is initialized once before the loop and overwritten with the model's predicted action each step:

```python
prev_state = torch.tensor([[0.0, -1.0, -1.0]], dtype=torch.float32)
…
for i in range(len(ep)):
    …
    batch = {…, "observation.state": prev_state, …}
    a = policy.select_action(mb)
    …
    btn_pred = 1.0 if float(a[0]) > 0.5 else 0.0
    prev_state = torch.tensor([[btn_pred, x_pred, y_pred]], dtype=torch.float32)
    policy._queues["action"].clear()
```

The paper's offline-evaluation description (`sec/benchmark.tex:121`) says:

> Offline Evaluation is performed as an open-loop assessment of the policy, measuring the policy's ability to replicate the ground-truth stepwise behavior without compounding errors. Given a static environment where screenshot, task and **oracle previous state are provided**, …

The released `h-siyuan/ScreenDrag` parquet's `observation.state` column happens to be exactly those oracle values — at frame `i` it holds the GT action at frame `i-1`, with the sentinel `[0, -1, -1]` at `i=0`:

```
frame 0: state=[0.0, -1.0, -1.0]   action=[1.0, 829.0, 697.5]
frame 1: state=[1.0, 829.0, 697.5]  action=[1.0, 840.5, 693.6]
frame 2: state=[1.0, 840.5, 693.6]  action=[1.0, 852.0, 689.7]
```

So I'm wondering whether the intended behavior is to read `observation.state` directly from the parquet at each step (matching paper text), or whether the current free-running rollout was intentional and the paper's wording is describing something slightly different. The two protocols give quite different numbers (details below); I'd love clarification on which one matches Tab 1.

### Per-step error pattern (captcha, 101 episodes)

For comparison, I ran the script as-is ("free-running") alongside a version that reads `observation.state` from the parquet ("oracle"):

| step | free-running mean err (px) | free-running within-20 px | oracle mean err (px) | oracle within-20 px |
|---:|---:|---:|---:|---:|
| 0 | 63.9 | 67 % | 81.1 | 66 % |
| 5 | 94.1 | 15 % | 18.2 | 64 % |
| 10 | 106.1 | 12 % | 22.6 | 48 % |
| 15 | 138.4 | 7 % | 19.2 | 64 % |
| 20 | 205.5 | 2 % | 25.0 | 48 % |

Free-running's mean error grows monotonically over the 21 steps (typical of compounding state errors), whereas the oracle setting stays roughly flat after step 0. The two starting points are similar, which suggests it's not the model itself behaving differently — only the state input changes.

### Headline numbers (3 seeds, all 5 domains, paper Tab 1 layout)

Same `showlab/ShowUI-pi` checkpoint, same `h-siyuan/ScreenDrag` test split, identical inference config in both runs — only the state source differs. Seeds 0, 1, 2 set via `torch.manual_seed`. CLI summary and the synced `aggregate.json` on S3 are byte-identical across all 3 seeds (verified).

**Endpoint Accuracy (↑, %, `endpoint_accuracy_best` = any-step within ε=20 px of GT endpoint):**

| Source | OS | PPT | Premiere | Captcha | Handwriting | Overall |
|---|---|---|---|---|---|---|
| Paper Tab 1 — ShowUI-π (450M) | 61.39 | 85.06 | 56.92 | 96.30 | 93.07 | **78.55** |
| Ours, oracle state (3 seeds, mean ± std) | 52.48 ± 2.62 | 86.14 ± 3.96 | 90.76 ± 1.14 | 69.31 ± 4.95 | 87.46 ± 2.86 | **77.23 ± 0.20** |
| Ours, free-running (current `main`, 3 seeds) | 21.78 | 46.53 | 45.54 | 16.83 | 39.60 | **34.06** |

5-dom mean with the patch is **77.23 ± 0.20 %**, within 1.3 pp of paper's 78.55 %. Without the patch, the same script averages 34 % — well under half of paper.

**Trajectory Error (↓, px, ATE = mean per-step L2 error):**

| Source | OS | PPT | Premiere | Captcha | Handwriting | Overall |
|---|---|---|---|---|---|---|
| Paper Tab 1 — ShowUI-π (450M) | 350.55 | 57.68 | 195.96 | 136.47 | 54.57 | **159.05** |
| Ours, oracle state (3 seeds, mean ± std) | 34.82 ± 0.67 | 57.10 ± 0.70 | 34.03 ± 1.30 | 22.58 ± 0.27 | 22.18 ± 0.10 | **34.14 ± 0.51** |

Our ATE is **~4.7× lower** than paper overall; PPT alone matches paper exactly (57.10 ± 0.70 vs 57.68). This is a separate question worth flagging — likely tied to the released parquet having 21 frames/episode vs the paper's described ~577-fps recordings (`benchmark.tex:70`), but I'd love confirmation.

### What I tried — minimal diff (against current `main`)

This is what I used for the "oracle" run above. Happy to revise or send a PR if it's the right direction — I just wanted to ask first.

```diff
@@ scripts/eval_dex.py: def run_episode(...)
     task_index = int(ep.iloc[0]["task_index"])
     task_text = tasks_by_id.get(task_index, "")

-    prev_state = torch.tensor([[0.0, -1.0, -1.0]], dtype=torch.float32)
     best_err = float("inf")
     final_err = float("inf")

     with torch.no_grad(), torch.autocast("cuda", dtype=torch.bfloat16):
         for i in range(len(ep)):
             row = ep.iloc[i]
             img = _decode_image(row["observation.images.screen"])

+            # Use the parquet's observation.state directly — at frame i it
+            # contains the GT action at frame i-1 (sentinel [0,-1,-1] at i=0).
+            gt_state = np.asarray(row["observation.state"], dtype=np.float64)
+            prev_state = torch.tensor([gt_state.tolist()], dtype=torch.float32)
+
             batch = {
                 "observation.images.screen": img,
                 "observation.state": prev_state,
                 "task": task_text,
             }
             mb = preprocessor(batch)
             act = policy.select_action(mb)
             act = postprocessor(act)
             a = act.numpy()[0]

             x_pred = float(np.clip(a[1], 0, 1919))
             y_pred = float(np.clip(a[2], 0, 1079))

             err = math.sqrt((x_pred - tgt_x) ** 2 + (y_pred - tgt_y) ** 2)
             if err < best_err:
                 best_err = err
             final_err = err

-            btn_pred = 1.0 if float(a[0]) > 0.5 else 0.0
-            prev_state = torch.tensor(
-                [[btn_pred, x_pred, y_pred]], dtype=torch.float32
-            )
-
-            # Force replanning at every step so state feedback actually matters.
             policy._queues["action"].clear()
```

### A secondary question — the Trajectory Error gap

The patch brings the 5-dom mean `endpoint_accuracy_best` within 1.3 pp of paper Tab 1, but per-domain rows diverge and **ATE is uniformly ~4.7× lower** than paper (see table above). Some observations:

- **PowerPoint ATE matches paper to within 1 px** (57.10 ± 0.70 vs 57.68). Every other domain is 2.5–10× lower.
- Captcha per-domain `endpoint_accuracy` is ~51 % vs paper 96.3 %. A threshold sweep on our captcha run hits ~94-96 % at **ε between 50 and 75 px** (script hard-codes ε = 20 px, and paper text says "*e.g.*, 20 pixels" — the "e.g." suggests 20 was an illustrative number).
- Both effects are consistent with paper having measured on a denser test set: the released `test/parquet/` averages ~20 frames/episode, while `benchmark.tex:70` describes the underlying data as ~577 frames/recording. Denser ground-truth waypoints → larger per-step displacement to compare against → larger ATE; and a less constrained acceptance region per step at the same ε would also account for the Endpoint Accuracy gap.

Two clarifying questions, if you can spare a moment:
1. Was Tab 1 measured on the released `test/parquet/` (≤ 21 frames/ep) or on the dense recordings?
2. What's the actual ε you used for Endpoint Accuracy in Tab 1?

Happy to spin this off as a separate issue if it muddies the state-handling fix.

### Reproduction

```bash
# Current main (free-running state):
PYTHONPATH=lerobot/src python scripts/eval_dex.py \
  --ckpt <local dir snapshot_download'd from showlab/ShowUI-pi> \
  --domain captcha --seed 0
# observed:  endpoint_accuracy ≈ 2-4 %, endpoint_accuracy_best ≈ 17-24 %, ATE ≈ 100-130 px

# After applying the diff above (oracle state):
# observed:  endpoint_accuracy ≈ 45-57 %, endpoint_accuracy_best ≈ 64-74 %, ATE ≈ 22-23 px

# Full 5-domain × 3-seed reproduction (across sm12/sm11/cloud-ru on our cluster):
# 5-dom mean endpoint_accuracy_best = 77.23 ± 0.20 %  (paper Tab 1: 78.55)
# 5-dom mean ATE                    = 34.14 ± 0.51 px (paper Tab 1: 159.05)
```

Thanks again for releasing all of this — let me know if I'm reading the protocol wrong and I'll close the issue.

**Comment by rakhimovv** [comment]
Follow-up on the secondary frame-density point in this issue — the same concern shows up much more sharply in Tab 2 (online), so I wanted to fold the evidence in here rather than open a third issue.

### Setup

I implemented the data-driven online protocol from `data.tex:40-67` from scratch (no released Tab 2 script that I could find). Same `showlab/ShowUI-pi` checkpoint, released `data/ScreenDrag/test/raw/` (505 episodes total, 101 per domain × 5 domains), ε = 20 px matcher, success = any predicted point within ε of the final GT waypoint. Seeds 0, 1, 2 via `torch.manual_seed`.

### Headline numbers (3 seeds, paper Tab 2 layout)

**Online Success Rate (↑, %):**

| Source | OS | PPT | Premiere | Captcha | Handwriting | Overall |
|---|---|---|---|---|---|---|
| Paper Tab 2 — ShowUI-π (450M) | 13.11 | 22.93 | 8.64 | 55.91 | 34.32 | **26.98** |
| Ours (3 seeds, mean ± std) | 0.99 ± 0.00 | 3.96 ± 0.99 | 4.29 ± 0.57 | 11.55 ± 1.51 | 3.30 ± 1.51 | **4.82 ± 0.69** |

That's a ~5.6× overall gap, with the largest absolute deltas on Captcha (~5×) and Handwriting (~10×) — exactly the two cells that dominate paper Tab 2's mean.

### Why I think this is the same concern as the Tab 1 ATE gap

The released `data/ScreenDrag/test/raw/` contains 10,289 PNG frames across 505 episodes (~20.4 frames per episode on average). `benchmark.tex:70` describes the underlying source recordings as ~577 frames at 60 fps. One root cause — Tab 1 and Tab 2 having been measured on the dense recordings while the released test set is downsampled — would explain both:

- **Tab 1 ATE** (per-step L2): sparser waypoints → larger per-step displacement to compare against → our ATE is ~4.7× *lower* than paper's, consistent with smaller inter-frame motion in the released sparse data.
- **Tab 2 online SR**: fewer intermediate matcher hits available → harder to chain together a successful rollout → our SR is ~5.6× *lower* than paper's.

Both directions are what you'd expect if the released `test/raw/` is the downsampled version of what Tab 1/Tab 2 were measured on.

### Two short follow-ups, if you have a moment

1. Was Tab 2 measured on the released `test/raw/` (≤ 21 frames/ep), or on the dense source recordings?
2. Would it be possible to release the dense recordings — even just one domain (e.g. captcha) — so the online column in Tab 2 becomes reproducible end-to-end?

If a Tab 2 reference script exists internally, that would also help me cross-check my from-scratch matcher against yours before drawing conclusions. Happy to share my `scripts/eval_dex_online.py` (290 LOC, mirrors `data.tex:40-67` as I read it) and the per-seed aggregates if useful — they're preserved on my side at `outputs/eval_dex_online_seed{0,1,2}/aggregate.json`.

Thanks again for any clarification.

**Comment by rakhimovv** [comment]
Related Tab 1 finding while we're here: the TEA formula in `benchmark.tex:138-142` doesn't reproduce the paper's Tab 1 number on the released checkpoint, no matter how I interpret it.

The LaTeX is `TEA = (1/T) Σ 1(â_t ∈ R_t)`, but `R_t` is only described as "the corresponding acceptance region" of radius ε (~20 px). The released `scripts/eval_dex.py` instead reports `endpoint_accuracy_best = 1(min_t ‖â_t − a_T^gt‖ ≤ 20 px)`: one boolean per episode, then averaged across episodes.

I tried four ways of reading the formula on `showlab/ShowUI-pi`, with `--state_mode oracle` (the fix from the original report above), 3 seeds, mean ± std over the 5 domains:

| Variant | Formula (ε = 20 px) | 3-seed 5-dom mean ± std |
|---|---|---:|
| A. text formula, per-step target | `(1/T) Σ 1(‖â_t − a_t^gt‖ ≤ ε)` | 47.90 ± 0.27 |
| B. text formula, final target | `(1/T) Σ 1(‖â_t − a_T^gt‖ ≤ ε)` | 11.06 ± 0.11 |
| C. single-endpoint (the commented-out form at `benchmark.tex:135-137`) | `1(‖â_T − a_T^gt‖ ≤ ε)` | 55.18 ± 2.70 |
| D. released `endpoint_accuracy_best` | `1(min_t ‖â_t − a_T^gt‖ ≤ ε)` | 77.23 ± 0.20 |

Paper Tab 1 reports 78.55. Only D lands within 1.3 pp; A/B/C are 23-67 pp off.

Which one did you actually use? If it's D, the LaTeX formula could be updated to match; if it's something else, the exact definition would be useful to know.
